File: android-app/app/src/main/java/com/kvnetflix/mobile/viewmodel/MyListViewModel.kt
package com.kvnetflix.mobile.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kvnetflix.mobile.data.model.Movie
import com.kvnetflix.mobile.data.repository.MovieRepository
import com.kvnetflix.mobile.data.repository.UserDataRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

data class MyListUiState(
    val savedMovies: List<Movie> = emptyList(),
    val watchHistory: List<Movie> = emptyList(),
    val exploreMovies: List<Movie> = emptyList(),
    val isLoadingExplore: Boolean = false
)

class MyListViewModel(application: Application) : AndroidViewModel(application) {

    private val userRepo = UserDataRepository(application)
    private val movieRepo = MovieRepository()
    private val _uiState = MutableStateFlow(MyListUiState())
    val uiState: StateFlow<MyListUiState> = _uiState

    init {
        viewModelScope.launch {
            userRepo.myList.collectLatest { list ->
                _uiState.value = _uiState.value.copy(savedMovies = list)
            }
        }
        viewModelScope.launch {
            userRepo.watchHistory.collectLatest { history ->
                _uiState.value = _uiState.value.copy(watchHistory = history)
            }
        }
        loadExplore()
    }

    fun loadExplore() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingExplore = true)
            try {
                val movies = movieRepo.exploreMovies()
                _uiState.value = _uiState.value.copy(
                    exploreMovies = movies,
                    isLoadingExplore = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoadingExplore = false)
            }
        }
    }

    fun removeFromMyList(slug: String) {
        viewModelScope.launch { userRepo.removeFromMyList(slug) }
    }

    fun removeFromHistory(slug: String) {
        viewModelScope.launch {
            userRepo.addToHistory(Movie(slug = slug, title = "", thumbnail = "", category = ""))
        }
    }
}

File: ios-app/KvNetflix/ViewModels/MyListViewModel.swift
import Foundation
import Combine

@MainActor
public final class MyListViewModel: ObservableObject {
    @Published public private(set) var savedMovies: [Movie] = []
    @Published public private(set) var watchHistory: [Movie] = []
    @Published public private(set) var exploreMovies: [Movie] = []
    @Published public private(set) var isLoadingExplore: Bool = false

    private let userRepo = UserDataRepository.shared
    private let movieRepo = MovieRepository.shared
    private var cancellables = Set<AnyCancellable>()

    public init() {
        setupObservers()
        loadExplore()
    }

    private func setupObservers() {
        userRepo.$myList
            .receive(on: DispatchQueue.main)
            .sink { [weak self] list in
                self?.savedMovies = list
            }
            .store(in: &cancellables)

        userRepo.$watchHistory
            .receive(on: DispatchQueue.main)
            .sink { [weak self] history in
                self?.watchHistory = history
            }
            .store(in: &cancellables)
    }

    public func loadExplore() {
        isLoadingExplore = true
        Task {
            do {
                let movies = try await movieRepo.exploreMovies()
                self.exploreMovies = movies
                self.isLoadingExplore = false
            } catch {
                self.isLoadingExplore = false
                print("[MyListViewModel] Failed to load explore recommendations: \(error)")
            }
        }
    }

    public func removeFromMyList(slug: String) {
        userRepo.removeFromMyList(slug: slug)
    }

    public func removeFromHistory(slug: String) {
        // Creates a placeholder or removes directly
        var updated = userRepo.watchHistory.filter { $0.slug != slug }
        // Directly update user repository watch history
        if let data = try? JSONEncoder().encode(updated) {
            UserDefaults.standard.set(data, forKey: "kv_user_watch_history")
        }
        userRepo.clearProgress(slug: slug)
    }

    public func refresh() {
        loadExplore()
        Task {
            await userRepo.syncWithRemote()
        }
    }
}
